Feature/usability: Allow input of precalculated time series and different SAM SI modules

Piranias opened the pull request:
Fix #313, #300, #285, #312, #165, #125

**Changes proposed in this pull request**:
- add new input parameters to main.apply_pvcompare()
- new function "add_pv_timeseries()" to pv_feedin.py

The following steps were realized, as well (required):
- [x] Update the CHANGELOG.md
- [x] Check if full simulation tests pass locally (`EXECUTE_TESTS_ON=master pytest`)

Also the following steps were realized (if applies):
- [x] Use in-line comments to explain your code
- [x] Write docstrings to your code
- [x] For new functionalities: Explain in readthedocs
- [x] Write test(s) for your new patch of code


*Please mark above checkboxes as following:*
- [ ] Open
- [x] Done

In case of an error due to linting, run `black . --exclude docs/` and push your changes.
Note that in case you do not fix a whole issue you should start your PR with `Address #xyz`.

<sub>*For more information on how to contribute check the [CONTRIBUTING.md](https://github.com/rl-institut/mvs_eland/blob/dev/CONTRIBUTING.md).*<sub>

Piranias committed:
allow own weather data

Piranias committed:
add_weather_file and add_sam_si_module

Piranias committed:
add_demand_timeseries

Piranias committed:
add_demand_timeseries

Piranias committed:
add docstrings for add_sam_si_module

Piranias committed:
add more description for main function

Piranias committed:
allow add pv timeseries

Piranias committed:
add docstring to new function

Piranias committed:
add changelog

Piranias committed:
apply black

Piranias committed:
add sam_si_module to nominal value

Piranias committed:
add tests for new features to test_main()

Piranias committed:
add predefined timeseries to test data

Piranias committed:
adjust tests

Piranias committed:
fix tests and docstrings

Piranias commented:
@SabineHaas , I will still add more informations about the features in the rtd. If you want you can start your review on the functionality and give me a feedback already

SabineHaas commented on pvcompare/main.py:
```diff
@@ -92,6 +97,37 @@ def apply_pvcompare(
+    add_electricity_demand: str or None
+        Path to precalculated hourly electricity demand time series for one year (or the same period
+        of a precalculated PV timeseries)
+        Note that that the demand is only considered is a column "Electricity demand" is added to
```
```suggestion
        Note that that the demand is only considered if a column "Electricity demand" is added to
```

SabineHaas commented on pvcompare/main.py:
```diff
@@ -92,6 +97,37 @@ def apply_pvcompare(
         absolute and relative will be overwritten with calculated time series of fixed thermal
         losses relative and absolute.
+    add_weather_data: str or None
+        Path to hourly csv weather time series with columns: [time, latitude, longitude
```
```suggestion
        Path to csv containing hourly weather time series with columns: [time, latitude, longitude
```

SabineHaas reviewed (commented):
Nice that you make this possible :)

Concerning weather data and pv time series I agree that it makes sense to introduce a parameter. 
For demand I think it would be more convenient to check whether the file named in `energyConsumption.csv` already exists (no calcluation) or not (demand time series is calculated and saved + file name saved to energyConsumption.csv).

This is how it is done for COPs of heat pumps already [around here](https://github.com/greco-project/pvcompare/blob/cd5516527a4b13e69e9f471ab1a2cfd25d298ea8/pvcompare/heat_pump_and_chiller.py#L400)

SabineHaas commented:
okay, I am fine with this - and I agree that it is more flexible. 
Could you adapt the process for COPs as well then? I think it should be consistent.

Piranias commented:
> okay, I am fine with this - and I agree that it is more flexible.
> Could you adapt the process for COPs as well then? I think it should be consistent.

I have come across the mechanism with the COPs before and I know that it's very interlaced. Therefore I'd prefer @MaGering to do this in another PR. I don't want to mess up your structure :)

Piranias committed:
add decription of new features in docs

Piranias committed:
adapt doctring

Piranias committed:
add energy system grapg to finish #165

Piranias committed:
Update pvcompare/main.py

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Update pvcompare/main.py

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Merge branch 'dev' into feature/usability

SabineHaas commented on docs/basic_usage.rst:
```diff
@@ -77,11 +86,62 @@ install the *cdsapi* package. `This page <https://cds.climate.copernicus.eu/api-
+E.g. {"cecmod":"Canadian_Solar_Inc__CS5P_220M"}.
 
+Note that the SI module is only considered if a module with the technology "SI" is provided in
+``user_inputs/mvs_inputs/pvcompare_inputs/pv_setup.csv``
```
```suggestion
``user_inputs_pvcompare_directory/pv_setup.csv``
```

SabineHaas commented on docs/basic_usage.rst:
```diff
@@ -77,11 +86,62 @@ install the *cdsapi* package. `This page <https://cds.climate.copernicus.eu/api-
+
+In order to add your own heat and/or electricity time series add the path to the file to ``add_electricity_demand`` or ``add_heat_demand``
+in :py:func:`~.main.apply_pvcompare`. Note that the respective demand is only considered, if a column "Electricity demand" or "Heat demand"
+exists in ``data/user_inputs/mvs_inputs/csv_elements/energyConsumption.csv``.
```
I'm not sure how I would provide the paths here in RTD. I have adapted the paths in the docstring in main.py (see PR), as there it is defined by the parameters `user_inputs_mvs_directory`, etc. 
In RTD I am not sure.. you can decide!

SabineHaas commented on docs/model_assumptions.rst:
```diff
@@ -98,6 +98,7 @@ unique models were developed for the CPV and PeroSi technologies. The next
 approaches.
 
+.. _si:
 1. SI
```
```suggestion

1. SI
```

SabineHaas commented on pvcompare/demand.py:
```diff
@@ -81,6 +83,12 @@ def calculate_load_profiles(
         Default: None.
     user_inputs_mvs_directory : str or None
         Path to mvs input directory. If None: DEFAULT_USER_INPUTS_MVS_DIRECTORY. Default: None.
+    add_electricity_demand: str
```
Default None for both

SabineHaas commented on pvcompare/main.py:
```diff
@@ -92,6 +97,37 @@ def apply_pvcompare(
         overwritten with calculated time series of COP and existing fixed thermal losses
         absolute and relative will be overwritten with calculated time series of fixed thermal
         losses relative and absolute.
+    add_weather_data: str or None
```
I have suggested adaptions to this docstring in a PR

Piranias replied:
which PR?

SabineHaas replied:
#322

SabineHaas commented on pvcompare/main.py:
```diff
@@ -130,39 +166,55 @@ def apply_pvcompare(
-        year=year,
-        normalization=True,
-    )
+    # check if add_pv_time_series is True
```
```suggestion
    # check if add_pv_time_series is provided
```

SabineHaas commented on pvcompare/main.py:
```diff
@@ -276,18 +330,33 @@ def apply_mvs(
     year = 2018
```
Do we actually need this part under `if __name__ == "__main__":`? I thought this was only for testing during development..
Accounts for all modules but examples, I think..

Piranias replied:
no, we do not need it and it should be deleted before the next release

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -96,7 +97,9 @@ def create_pv_components(
         returned. Default: True.
-
+    add_sam_si_module: dict
+        with library (’CECMod’  or "SandiaMod") as key and module name as value.
```
```suggestion
        Dictionary with library ("CECMod"  or "SandiaMod") as key and module name as value.
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -96,7 +97,9 @@ def create_pv_components(
-
+    add_sam_si_module: dict
+        with library (’CECMod’  or "SandiaMod") as key and module name as value.
+        E.g. {"cecmod":'Canadian_Solar_Inc__CS5P_220M'}
```
```suggestion
        E.g. {"cecmod": "Canadian_Solar_Inc__CS5P_220M"}
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -311,6 +316,9 @@ def set_up_system(technology, surface_azimuth, surface_tilt):
         surface azimuth of the module
     surface_tilt: float
         surface tilt of the module
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
```
```suggestion
        Dictionary with library ("CECMod"  or "SandiaMod") as key and module name as value.
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -311,6 +316,9 @@ def set_up_system(technology, surface_azimuth, surface_tilt):
     surface_tilt: float
         surface tilt of the module
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
+        E.g. {"cecmod":'Canadian_Solar_Inc__CS5P_220M'}.
```
```suggestion
        E.g. {"cecmod": "Canadian_Solar_Inc__CS5P_220M"}.
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -311,6 +316,9 @@ def set_up_system(technology, surface_azimuth, surface_tilt):
         surface tilt of the module
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
+        E.g. {"cecmod":'Canadian_Solar_Inc__CS5P_220M'}.
+        Note that the SI module is only considered if there is the technology "SI" in
```
something missing here?

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -392,14 +412,21 @@ def create_si_time_series(
         If True: Time series is normalized. Otherwise absolute time series is
         returned.
+    add_sam_si_module: dict
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
```
```suggestion
        Dictionary with library ("CECMod"  or "SandiaMod") as key and module name as value.
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -392,14 +412,21 @@ def create_si_time_series(
         returned.
+    add_sam_si_module: dict
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
+        E.g. {"cecmod":'Canadian_Solar_Inc__CS5P_220M'}.
```
```suggestion
        E.g. {"cecmod": "Canadian_Solar_Inc__CS5P_220M"}.
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -392,14 +412,21 @@ def create_si_time_series(
+    add_sam_si_module: dict
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
+        E.g. {"cecmod":'Canadian_Solar_Inc__CS5P_220M'}.
+        Note that the SI module is only considered if there is the technology "SI" in
```
also something missing

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -605,6 +634,10 @@ def nominal_values_pv(technology, area, surface_azimuth, surface_tilt, psi_type)
     surface_tilt: float
         surface tilt of the modules
+    add_sam_si_module: dict or None
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
```
```suggestion
        Dictionary with library ("CECMod"  or "SandiaMod") as key and module name as value.
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -605,6 +634,10 @@ def nominal_values_pv(technology, area, surface_azimuth, surface_tilt, psi_type)
         surface tilt of the modules
+    add_sam_si_module: dict or None
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
+        E.g. {"cecmod":'Canadian_Solar_Inc__CS5P_220M'}.
```
```suggestion
        E.g. {"cecmod": "Canadian_Solar_Inc__CS5P_220M"}.
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -605,6 +634,10 @@ def nominal_values_pv(technology, area, surface_azimuth, surface_tilt, psi_type)
+    add_sam_si_module: dict or None
+        Dictionary with library (’CECMod’  or "SandiaMod") as key and module name as value.
+        E.g. {"cecmod":'Canadian_Solar_Inc__CS5P_220M'}.
+        Note that the SI module is only considered if there is the technology "SI" in
```
something missing

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -695,3 +729,78 @@ def get_peak(technology, module_parameters_1, module_parameters_2):
+    user_inputs_mvs_directory,
+    user_inputs_pvcompare_directory,
+):
+    """
```
```suggestion
    r"""
    Adds PV time series of user to the simulations inputs.
    
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -695,3 +729,78 @@ def get_peak(technology, module_parameters_1, module_parameters_2):
+    user_inputs_pvcompare_directory,
+):
+    """
+    This function calculates the maximal capacity and inserts the time series filename
```
```suggestion
    This function calculates the maximaum capacity and inserts the time series filename
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -695,3 +729,78 @@ def get_peak(technology, module_parameters_1, module_parameters_2):
+):
+    """
+    This function calculates the maximal capacity and inserts the time series filename
+    and the maximum capacity into 'user_inputs/mvs_inputs/csv_elements/energyProduction.csv'.
```
```suggestion
    and the maximum capacity into 'user_inputs_mvs_directory/csv_elements/energyProduction.csv'.
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -695,3 +729,78 @@ def get_peak(technology, module_parameters_1, module_parameters_2):
+         When providing your own time series, overwrite_pv_parameters should be
+        set to false. When add_pv_timeseries is used, the pv_setup.csv is disregarded.
+    storeys: int
+        number of storeys for which the demand is calculated.
```
please adapt description

Piranias replied:
you mean the description if the add_pv_timeseries parameter? I updated it to the docs version

SabineHaas replied:
Sorry, I meant the description of the above line, of `storeys`

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -695,3 +729,78 @@ def get_peak(technology, module_parameters_1, module_parameters_2):
+        as  user_inputs_pvcompare_directory.
+        Default: None.
+
+    Return
```
```suggestion
    Returns
```

SabineHaas commented on pvcompare/pv_feedin.py:
```diff
@@ -695,3 +729,78 @@ def get_peak(technology, module_parameters_1, module_parameters_2):
+        if not os.path.isfile(add_pv_timeseries[key]["filename"]):
+            logging.error(
+                "The PV time series you have specified does not exist. "
+                "Please check your input or set 'add_pv_timeseries' to None "
```
```suggestion
                "Please check your input or set `add_pv_timeseries` to None "
```

SabineHaas commented on tests/test_main.py:
```diff
@@ -77,6 +83,152 @@ def test_apply_pvcompare(self):
         )
 
+    def test_apply_pvcompare_add_weather_file(self):
+        # delete file
```
does this affect any other tests?

Piranias replied:
no, it doesn't

SabineHaas commented on pvcompare/demand.py:
```diff
@@ -105,28 +113,66 @@ def calculate_load_profiles(
+                            user_inputs_mvs_directory=user_inputs_mvs_directory,
+                        )
+                    else:
+                        logging.warning(
```
Careful: is the demand calculated here, if the file does not exist?

Piranias replied:
Yes. and a warning is given. You want to raise an error instead?

SabineHaas commented on pvcompare/demand.py:
```diff
@@ -105,28 +113,66 @@ def calculate_load_profiles(
+                        )
+                    else:
+                        logging.warning(
+                            "The electricity demand time series"
```
same as above: demand is only calcualted if add_electricity_demand is not None

SabineHaas reviewed (changes requested):
It's important that we first merge #309, otherwise @Stefanie08 could get nasty merge conflicts which might be difficult for her to solve. 

- I've found some strange dots here: https://pvcompare.readthedocs.io/en/latest/model_assumptions.html#id18 maybe you can fix this as well?
- It would be nice if you add assertion statements to your tests
- You could also test the cases that one wants to provide own time series but the file does not exist (and is then calculated by pvcompare) or provides a wrongly formatted dict for PV

SabineHaas commented on docs/basic_usage.rst:
```diff
@@ -12,6 +12,15 @@ Run a simulation
 
+Here's an example energy system graph of a sector coupled scenario of the simulation provided in ``examples/run_pvcompare_example_sector_coupling.py``.
+
+.. figure:: ./images/basic_usage_energy_system_graph.png
```
I think the figure is not uploaded yet

SabineHaas commented:
> It's important that we first merge #309, otherwise @Stefanie08 could get nasty merge conflicts which might be difficult for her to solve.

@Piranias I have thought about the merge again: I would do it after the release. The reason for this is that I would reference pvcompare with the release tag in our working paper. Of course, the changes in this PR should not change any results, but I still prefer to state the version of pvcompare that we have used for the simulations. 

I have fixed the bug for which we could not do the release - maybe we can even finish it this week. If not, you can still finalize this PR and I can merge it after the release is done.

Piranias committed:
Update CHANGELOG.md

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Update docs/basic_usage.rst

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Update docs/basic_usage.rst

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias commented on docs/basic_usage.rst:
```diff
@@ -77,11 +86,62 @@ install the *cdsapi* package. `This page <https://cds.climate.copernicus.eu/api-
+
+In oder to provide your own weather data, you need to insert the path
+to your weather file to the parameter ``add_weather_data`` in the :py:func:`~.main.apply_pvcompare` function. The file
+should contain an hourly time series in *csv* format with the columns: [time, latitude, longitude
```
i think the hourly is important..

SabineHaas replied:
I have moved this part after the brackets [].
By the way, in the files tab you can add suggestions to the batch and commit them in one commit ;)

Piranias committed:
Update docs/basic_usage.rst

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Update docs/basic_usage.rst

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Update pvcompare/pv_feedin.py

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Update pvcompare/pv_feedin.py

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Merge branch 'dev' into feature/usability

Piranias committed:
fix heat demand if file does not exist

Piranias committed:
add energy system graph

Piranias committed:
update name of graph

Piranias committed:
Update docs/basic_usage.rst

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Update docs/basic_usage.rst

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
Apply suggestions from code review

Co-authored-by: Sabine Haas <sabine.haas@rl-institut.de>

Piranias committed:
update docstrings

Piranias committed:
remove dots

Piranias committed:
fix raise error and further tests

Piranias commented:
> It's important that we first merge #309, otherwise @Stefanie08 could get nasty merge conflicts which might be difficult for her to solve.
> 
>     * I've found some strange dots here: https://pvcompare.readthedocs.io/en/latest/model_assumptions.html#id18 maybe you can fix this as well?
> 
>     * It would be nice if you add assertion statements to your tests
> 
>     * You could also test the cases that one wants to provide own time series but the file does not exist (and is then calculated by pvcompare) or provides a wrongly formatted dict for PV

i added the tests and updated all other changes. Except of the assertion statements, which i found less important right now, since we dont have any for the other tests as well..

Piranias committed:
Merge branch 'dev' into feature/usability

Piranias committed:
apply black
